be/src/exec/pipeline/exchange/exchange_sink_operator.cpp

```diff
@@ -195,31 +195,37 @@ Status ExchangeSinkOperator::Channel::send_chunk_request(PTransmitChunkParamsPtr
     return Status::OK();
 }
 
-Status ExchangeSinkOperator::Channel::_close_internal() {
+Status ExchangeSinkOperator::Channel::_close_internal(RuntimeState* state, FragmentContext* fragment_ctx) {
```

Comment:
Needn't pass FragmentContext here. we could check `is_canceled ` from RuntimeState.

```
    bool is_cancelled() const { return _is_cancelled; }
    void set_is_cancelled(bool v) { _is_cancelled = v; }
```

Reply:
`set_is_cancelled()` is only called in the calling sequence: `PInternalServiceImpl::cancel_plan_fragment -> FragmentMgr::cancel ->  FragmentExecState::cancel -> PlanFragmentExecutor::cancel -> RuntimeState::set_is_cancelled`, but `FragmentMgr::cancel` is only called by `PInternalServiceImpl::cancel_plan_fragment` when it is not pipeline.

```C++
template <typename T>
void PInternalServiceImpl<T>::cancel_plan_fragment(google::protobuf::RpcController* cntl_base,
                                                   const PCancelPlanFragmentRequest* request,
                                                   PCancelPlanFragmentResult* result, google::protobuf::Closure* done) {
    if (request->has_is_pipeline() && request->is_pipeline()) {
        // ...
    } else {
        if (request->has_cancel_reason()) {
            st = _exec_env->fragment_mgr()->cancel(tid, request->cancel_reason());
        } else {
            LOG(INFO) << "cancel framgent, fragment_instance_id=" << print_id(tid);
            st = _exec_env->fragment_mgr()->cancel(tid);
        }
        if (!st.ok()) {
            LOG(WARNING) << "cancel plan fragment failed, errmsg=" << st.get_error_msg();
        }
    }
```

Reply:
In the internal of pipeline, `FragmentContext::_cancel_flag` is used as the cancel flag of a fragment. 
Except FE calling RPC `PInternalServiceImpl::cancel_plan_fragment`, the fragment may be cancelled by itself in `GlobalDriverDispatcher`, when `push_chunk()` and `pull_chunk()` return error.